Release 0.7.0 (#302)
kuzudb/kuzu-docs: src/content/docs/extensions/json.mdx

  </iframe>
</div>

### Example dataset

Let's look at an example dataset to demonstrate how the JSON extension can be used.

> We have 3 JSON files that contain information about patients and their medical conditions. The

Author: `person.json` should be renamed to `patient.json` in this file.

Reviewer: Ah thanks. Will fix soon.

RETURN to_json('kuzu');
```
```json
"kuzu"
```

Example 2:

```cypher
RETURN to_json([1,2,3]);
```
```json
[1,2,3]
```

Example 3:

```cypher
RETURN to_json('{"name": "Alicia", "age": 28}');
```
```json
{"name":"Alicia","age":28}
```

### `array_to_json`

Signature: `ARRAY -> JSON`

Alias for `to_json` that only accepts `ARRAY`.

### `row_to_json`

Signature: `LIST -> JSON`

Alias for `to_json` that only accepts `LIST`.

### `cast(ANY AS JSON)`

Signature: `ANY -> JSON`

Syntax sugar for `to_json(any) -> JSON`,
`cast` can cast `ANY` type to be some type other than `JSON` as well,
Read the instruction of `Casting` for more details.

Example:

```cypher
RETURN cast('{"name": "Alicia", "age": 28}' AS JSON);
```
```json
{"name":"Alicia","age":28}
```

### `json_object`

Signature: `STRING, ANY -> JSON object`

`json_object([key, value, ...])`

Create a JSON object from any number of `key`, `value` pairs.

Example 1:

```cypher
RETURN json_object("name","Alicia");
```
```json
{"name":"Alicia"}
```
Example 2:
```cypher
RETURN json_object("name","Alicia","age",28);
```
```json
{"name":"Alicia","age":28}
```

### `json_array`

Signature: `ARRAY -> JSON ARRAY`

`json_array([any, ...])`

> Create a JSON array from `any` number of values.

Author: ```suggestion
Create an array of JSON objects from **any** number of values. Each value is converted into a JSON object.
```

Example:

```cypher
RETURN json_array(42, 'kuzu', NULL);
```
```json
[42,"kuzu",null]
```

### `json_merge_patch`

Signature: `JSON, JSON -> JSON`

`json_merge_patch(json, json)`

Merges TWO `JSON` documents. Applies [RFC 7386](https://datatracker.ietf.org/doc/html/rfc7386)

Example 1:

```cypher
RETURN json_merge_patch('{"name": "Alicia"}', '{"age": 28}');
```
```json
{"name":"Alicia","age":28}
```
Example 2:
```cypher
RETURN json_merge_patch('3', NULL);
```
```json

> ```

Author: Remove empty json code block here.

Signatures: `JSON, STRING -> JSON`, `JSON, INTEGER -> JSON`, `JSON, LIST -> LIST of JSON`

`json_extract(json, path)`

Extracts `JSON` from `json` at the given `path`. 
`path` is a `STRING` delimited by `'/'`.
Integers may also be used to represent the path, which represents the index if the input is a JSON array.
If `path` is a `LIST` of paths, the result will also be a `LIST`.

If the path does not exist, this function returns an empty `JSON` document.

Examples:

In this case, we provide a single item as the `path`.

```cypher
RETURN json_extract('{"Software": {"Database": ["duck", "kuzu"]}}', 'Software/Database/1');
```
```json
"kuzu"
```

Here, we provide a `LIST` of paths:

```cypher
RETURN json_extract('{"Software": {"Database": ["duck", "kuzu"]}}', ['Software/Database/1', 'Software/Database/0']);
```
```json
["kuzu","duck"]
```

This example provides the path as an integer that represents the index of the item in the JSON array
we want to extract:

```cypher
RETURN json_extract('[1, 2, 42]', 2)
```
```json
42
```

Extracting from a nonexistent path results in empty `JSON` document:

```cypher
RETURN json_extract('{"Software": {"Database": ["duck", "kuzu"]}}');
```

> ```json

Author: Remove empty json code block here.

```

### `json_array_length`

Signature: `JSON -> UINT32`

`json_array_length(json[])`

> If the `json` is an JSON array, returns its length. Otherwise returns `0`.

Author: ```suggestion
If the `json` is an JSON array, return its length. Otherwise return `0`.
```

Example 1:

```cypher
return json_array_length('["1", "1", "4", null]');
```
```json
4
```
Example 2:
```cypher
return json_array_length('{"kuzu": ["1", "1", "4", null]}');
```
```json
0
```

> (this is not an JSON array)

Author: ```suggestion
Return 0 here as the input is not a JSON array.
```

### `json_contains`

Signature: `JSON, JSON -> BOOL`

`json_contains(json_haystack, json_needle)`

Returns `True` if `json_needle` is contained in `json_haystack`. Both parameters are of `JSON` type, but `json_needle` can also be a numeric value or a `string`, however the `string` must be wrapped in double quotes.

Example 1:

```cypher
RETURN JSON_CONTAINS('{"name": "Alicia"}', '"Alicia"');
```
```json
True
```
Example 2:
```cypher

> RETURN JSON_CONTAINS('{"age": 28}', '28');

Author: Can we also document that `'` is necessary here in the second parameter, or explain why that is necessary? I don't think this is intuitive for users to figure out.

```
```json
True
```

### `json_keys`

Signature: `JSON -> STRING[]`

`json_keys(json)`

> Gives the keys of the root `json` object. If the root is not an JSON object, returns an empty `list`.

Author: ```suggestion
Return keys of the root `json` object. If the root is not an JSON object, return an empty `list`.
```

Examples:

```cypher
return json_keys('{ "family": "anatidae", "species": [ "duck", "goose", "swan", null ] }');
```
```
[family,species]
```

### `json_structure`

Signature: `JSON -> STRING`

`json_structure(json)`

> Returns the structure of the `json` in Kuzu type notation. Will give smaller integer widths for smaller numbers.

Author: ```suggestion
Returns the structure of the `json` in Kùzu type notation. Integer widths are automatically inferred from input values.
```

Example:
```cypher
RETURN json_structure('[{"a": -1, "b": [1000, 2000, 3000]}, {"a": 2, "c": "hi"}]');
```
```
STRUCT(a INT16, b UINT16[], c STRING)[]
```

### `json_valid`

Signature: `JSON -> BOOL`

`json_valid(json)`

Determines whether or not the provided `json` is valid JSON.

Example 1:

```cypher
RETURN json_valid('{"name": "Alicia", "age": 28}');
```
```json
True
```
Example 2:
```cypher
RETURN json_valid('"name": "Alicia", "age": 28');
```
```json
False
```

### `json`

Signature: `JSON -> JSON`

`json(json)`

> Parses and minifies the `json.

Author: ```suggestion
Parses and minifies the `json`.
```

`json_extract(json, path)`

Extracts `JSON` from `json` at the given `path`. 
`path` is a `STRING` delimited by `'/'`.
Integers may also be used to represent the path, which represents the index if the input is a JSON array.
If `path` is a `LIST` of paths, the result will also be a `LIST`.

If the path does not exist, this function returns an empty `JSON` document.

Examples:

In this case, we provide a single item as the `path`.

```cypher
RETURN json_extract('{"Software": {"Database": ["duck", "kuzu"]}}', 'Software/Database/1') AS extracted;
```
```
┌───────────┐
│ extracted │
│ json      │
├───────────┤
│ "kuzu"    │
└───────────┘
```

Here, we provide a `LIST` of paths:

```cypher
RETURN json_extract('{"Software": {"Database": ["duck", "kuzu"]}}', ['Software/Database/1', 'Software/Database/0']) AS extracted;
```
```
┌─────────────────┐
│ extracted       │
│ json[]          │
├─────────────────┤
│ ["kuzu","duck"] │
└─────────────────┘
```

This example provides the path as an integer that represents the index of the item in the JSON array
we want to extract:

```cypher
RETURN json_extract('[1, 2, 42]', 2) AS nums;
```

> ```json

Reviewer: ```suggestion
```
```
